### PR TITLE
Initialization manager changes schema by changelog

### PR DESCRIPTION
**Pull request by Simcsa**
- initialization manager in DatabaseManagerBlImpl changes database version by changelog file in resources. It searches for the actual db version in the changelog file and executes all the commands of the newer version.
- DBVersion bean created for saving db versions with their commands
- tests to ensure that parsing of the changelog file works correctly were created. Test files created in resources in test folder.

### CODE REVIEW

**Comment by stavamichal** [comment]
This pull request need to be rebased on master before merging please.

Thanks.

File: perun-beans/src/main/java/cz/metacentrum/perun/core/api/DBVersion.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+import java.util.List;
+
+/**
+ * Object for representation of database version.
+ *
+ * @author Simona Kruppova 410315
+ * @author Oliver Mrazik 410035
+ */
+public class DBVersion extends Auditable {
```
Comment (stavamichal) [inline perun-beans/src/main/java/cz/metacentrum/perun/core/api/DBVersion.java]: This object not extended PerunBean and also not extended Auditable. Please remove extends part.

File: perun-beans/src/main/java/cz/metacentrum/perun/core/api/DBVersion.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+	@Override
+	public String toString() {
+		StringBuilder str = new StringBuilder();
+
+		return str.append(getClass().getSimpleName()).append(":[id='").append(getId()).append("', version='")
+				.append(version).append("', commands='").append(commands).append("']").toString();
+	}
+
+	public String serializeToString() {
```
Comment (stavamichal) [inline perun-beans/src/main/java/cz/metacentrum/perun/core/api/DBVersion.java]: Serialization is not needed if object not extends Auditable.

File: perun-beans/src/main/java/cz/metacentrum/perun/core/api/DBVersion.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * Object for representation of database version.
+ *
+ * @author Simona Kruppova 410315
+ * @author Oliver Mrazik 410035
+ */
+public class DBVersion {
+	String version;
+	List<String> commands;
```
Comment (stavamichal) [inline perun-beans/src/main/java/cz/metacentrum/perun/core/api/DBVersion.java]: It is good practice to use private before variables in objects. For purpose of setting and getting them there are public methods.

File: perun-core/src/main/java/cz/metacentrum/perun/core/bl/DatabaseManagerBl.java [hunk 4]
```diff
@@ -12,30 +14,49 @@
...
+	/**
+	 * Parse all database versions from DB changelog file and create list of DBVersion objects.
+	 *
+	 * @param currentDBVersion current DB version
+	 * @param fileName DB changelog file name, file should be in resources
+	 *
+	 * @return list of DBVersion objects
+	 *
+	 * @throws InternalErrorException
+	 */
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/bl/DatabaseManagerBl.java]: Write what happens if changelog file not exists or it is empty | wrong format.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/DatabaseManagerImpl.java [hunk 5]
```diff
@@ -73,4 +80,119 @@ public String getDatabaseInformation(PerunSession sess) throws InternalErrorExce
...
+
+	public List<DBVersion> getChangelogVersions(String currentDBVersion, String fileName) throws InternalErrorException {
+
+		Pattern versionPattern = Pattern.compile("^[1-9][0-9]*[.][0-9]+[.][0-9]+");
+		Pattern commentPattern = Pattern.compile("^--.*");
+
+		List<DBVersion> versions = new ArrayList<>();
+		boolean versionFound = false;
+
+		try(BufferedReader br = new BufferedReader(new InputStreamReader(ClassLoader.getSystemResourceAsStream(fileName)))) {
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/DatabaseManagerImpl.java]: Somewhere in code should be closing of stream (or file) readers.

File: perun-core/src/main/java/cz/metacentrum/perun/core/implApi/DatabaseManagerImplApi.java [hunk 6]
```diff
@@ -12,30 +14,60 @@
...
+	 *
+	 * @param currentDBVersion current DB version
+	 * @param fileName DB changelog file name, file should be in resources
+	 *
+	 * @throws InternalErrorException
+	 */
+	void updateDatabaseVersion(String currentDBVersion, String fileName) throws InternalErrorException;
+
+	/**
+	 * Parse all database versions from DB changelog file and create list of DBVersion objects.
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/implApi/DatabaseManagerImplApi.java]: Please add information (to all javadocs about this method) that it create list of DBVersion from currentDBVersion to now (without currentDBVersion itself).

**Comment by stavamichal** [comment]
Maybe there should be some method or test for verification all changelog files. Something what is able to check if versions are increasing so there is nothing like:
3.1.15
3.1.12
3.1.14

And everything is parsable without problem before even start db initialization (update). This is just a tip, try to think about it. Thank you.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/DatabaseManagerImpl.java (hunk 5 above)
Comment (Simcsa) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/DatabaseManagerImpl.java]: There is the try-with-resources statement used, it should ensure that the resource (or stream) is closed properly.

Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/DatabaseManagerImpl.java]: Sorry, you are right, this is new syntax for me.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/DatabaseManagerImpl.java [hunk 7]
```diff
@@ -73,4 +81,105 @@ public String getDatabaseInformation(PerunSession sess) throws InternalErrorExce
...
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/DatabaseManagerImpl.java]: After testing on devel I found a problem with ClassLoader in WAR file. In the tests there is just jar file which has no problem with defined classloader.
In WAR file need to be specific which ClassLoader will be used.
Use this instead:
Thread.currentThread().getContextClassLoader().getResourceAsStream(fileName)

**Comment by stavamichal** [comment]
And also please rebase it on actual master. Thank you.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/DatabaseManagerImpl.java (hunk 7 above)
Comment (Simcsa) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/DatabaseManagerImpl.java]: Done.

**Comment by zlamalp** [comment]
@stavamichal has own updated version of this feature, which will be eventually merged. Your work is included in it. Closing this pull-request. Thanks.

**Pull request closed without merging**